src/PowerShellEditorServices.Engine/Services/TextDocument/Handlers/CompletionHandler.cs

```diff
@@ -0,0 +1,320 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Management.Automation;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.PowerShell.EditorServices.Symbols;
+using OmniSharp.Extensions.LanguageServer.Protocol.Client.Capabilities;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+
+namespace Microsoft.PowerShell.EditorServices.TextDocument
+{
+    internal class CompletionHandler : ICompletionHandler, ICompletionResolveHandler
+    {
+        const int DefaultWaitTimeoutMilliseconds = 5000;
+        private readonly CompletionItem[] s_emptyCompletionResult = new CompletionItem[0];
+
+        private readonly ILogger _logger;
+        private readonly PowerShellContextService _powerShellContextService;
+        private readonly WorkspaceService _workspaceService;
+
+        private CompletionResults _mostRecentCompletions;
+
+        private int _mostRecentRequestLine;
+
+        private int _mostRecentRequestOffest;
+
+        private string _mostRecentRequestFile;
+
+        private CompletionCapability _capability;
+
+        public CompletionHandler(
+            ILoggerFactory factory,
+            PowerShellContextService powerShellContextService,
+            WorkspaceService workspaceService)
+        {
+            _logger = factory.CreateLogger<CompletionHandler>();
+            _powerShellContextService = powerShellContextService;
+            _workspaceService = workspaceService;
+        }
+
+        public CompletionRegistrationOptions GetRegistrationOptions()
+        {
+            return new CompletionRegistrationOptions
+            {
+                DocumentSelector = new DocumentSelector(new DocumentFilter { Pattern = "**/*.ps*1" }),
```

Comment:
Does this effect untitled files?

Reply:
Good catch, I can change this to apply to only powershell files